src/Common/tests/System/IO/CallTrackingStream.cs

```diff
@@ -0,0 +1,221 @@
+        {
+            UpdateCallCount();
+            ReadBuffer = buffer;
+            ReadOffset = offset;
+            ReadCount = count;
+            return Inner.Read(buffer, offset, count);
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            UpdateCallCount();
+            ReadAsyncBuffer = buffer;
+            ReadAsyncOffset = offset;
+            ReadAsyncCount = count;
+            ReadAsyncCancellationToken = cancellationToken;
+            return Inner.ReadAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override int ReadByte()
+        {
+            UpdateCallCount();
+            return Inner.ReadByte();
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            UpdateCallCount();
+            SeekOffset = offset;
+            SeekOrigin = origin;
+            return Inner.Seek(offset, origin);
+        }
+
+        public override void SetLength(long value)
+        {
+            UpdateCallCount();
+            SetLengthValue = value;
+            Inner.SetLength(value);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            UpdateCallCount();
+            WriteBuffer = buffer;
+            WriteOffset = offset;
+            WriteCount = count;
+            Inner.Write(buffer, offset, count);
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            UpdateCallCount();
+            WriteAsyncBuffer = buffer;
+            WriteAsyncOffset = offset;
+            WriteAsyncCount = count;
+            WriteAsyncCancellationToken = cancellationToken;
+            return Inner.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override void WriteByte(byte value)
+        {
+            UpdateCallCount();
+            WriteByteValue = value;
+            Inner.WriteByte(value);
+        }
+
+        // Bookkeeping logic
+
+        public int TimesCalled(string member) => _callCounts.GetOrDefault(member);
+
+        // [CallerMemberName] causes the member parameter to be set to the name
+        // of the calling member if not specified, e.g. calling this method
+        // from SetLength would pass in member with a value of "SetLength"
+        private T Read<T>(T property, [CallerMemberName] string member = null)
+        {
+            UpdateCallCount(member);
+            return property;
+        }
+
+        private void Update(Action setter, [CallerMemberName] string member = null)
+        {
+            UpdateCallCount(member);
+            setter();
+        }
+
+        private void UpdateCallCount([CallerMemberName] string member = null)
+        {
+            _callCounts[member] = _callCounts.GetOrDefault(member) + 1;
```

Comment:
This could be ```_callCounts[member] = TimesCalled(member);```, at which point there's only one call site to GetOrDefault, and you can delete the whole DictionaryExtensions class and just put the logic from GetOrDefault into TimesCalled.